py3status/modules/imap.py

```diff
@@ -65,61 +73,163 @@ class Meta:
+            response = socket.read(4096).decode('ascii')
+            if not response.lower().startswith('+ idling'):
+                raise IdleException(response)
+
+            # wait for IDLE to return
+            socket.setblocking(0)  # so we can timeout if IDLE doesn't return soon enough
+            if self.cache_timeout > 0:
+                ready = select.select([socket], [], [], self.cache_timeout)
+            else:
+                ready = select.select([socket], [], [])
+            if ready[0]:
+                # receive list of messages (we don't care what has changed, that
+                # gets checked in _get_mail_count() )
+                socket.read(4096)
+            socket.setblocking(1)
+
+        except IdleException as e:
+            raise imaplib.IMAP4.error("While initializing IDLE: " + str(e))
+        finally:
+            socket.write(b'DONE\r\n')  # important!
+            response = socket.read(4096).decode(encoding='ascii')
+            expected_response = (command_tag + b' OK Idle completed').decode(encoding='ascii')
+            if not response.lower().startswith(expected_response.lower()):
+                raise imaplib.IMAP4.abort("While terminating IDLE: " + response)
 
-            if self.security == "ssl":
-                connection = self._connection_ssl()
-            elif self.security == "starttls":
-                connection = self._connection_starttls()
-
-            connection.login(self.user, self.password)
-
-            for directory in directories:
-                connection.select(directory)
-                unseen_response = connection.search(None, self.criterion)
-                mails = unseen_response[1][0].split()
-                mail_count += len(mails)
-
-            connection.close()
-            return mail_count
-        except:
-            return None
+    def _get_mail_count(self):
+        try:
+            while True:
+                if self.connection is None:
+                    self._connect()
+                if self.connection.state is 'NONAUTH':
+                    self.connection.login(self.user, self.password)
+
+                tmp_mail_count = 0
+                directories = self.mailbox.split(',')
+
+                for directory in directories:
+                    self.connection.select(directory)
+                    unseen_response = self.connection.search(None, self.criterion)
+                    mails = unseen_response[1][0].split()
+                    tmp_mail_count += len(mails)
+
+                self.mail_count = tmp_mail_count
+
+                if self.use_idle:
+                    self.py3.update()
+                    self._idle()
+                else:
+                    return
+        except (socket_error, imaplib.IMAP4.abort, imaplib.IMAP4.readonly) as e:
+            self.py3.log("Recoverable error - " + str(e), level=self.py3.LOG_WARNING)
+            self._disconnect()
+            self.mail_count = None
```

Comment:
I just started using `imap`... Keep an eye on this line. It's causing `imap` to disappear occasionally... Maybe we should not hide it... Leave it alone as a cache. Lot of changes in this PR. :fearful:

Reply:
good idea. 

> Lot of changes in this PR

yeah; but it's not really something one can split up into multiple prs. :|